lookout/style/format/tests/test_quality_report.py

```diff
@@ -0,0 +1,155 @@
+from lookout.style.format.quality_report import quality_report
+
+
+PARENT = Path(__file__).parent.resolve()
+BASE_DIR = str(PARENT)
+FS = tempfile.TemporaryDirectory(dir=BASE_DIR)
+
+
+def tearDownModule():
+    """Required to delete trained rules model."""
+    FS.cleanup()
+
+
+class Capturing(list):
+    def __enter__(self):
+        self._stdout = sys.stdout
+        sys.stdout = self._stringio = io.StringIO()
+        return self
+
+    def __exit__(self, *args):
+        self.extend(self._stringio.getvalue().splitlines())
+        del self._stringio    # free up some memory
+        sys.stdout = self._stdout
+
+
+@unittest.skipUnless(os.getenv("LONG_TESTS", False),
+                     "Time-consuming tests are skipped by default.")
+class QualityReportTests(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        """Create temporary directory with jquery repository."""
+        cls.jquery_dir = str(PARENT / "jquery")
+        # str() is needed for Python 3.5
+        with tarfile.open(str(PARENT / "jquery.tar.xz")) as tar:
+            tar.extractall(path=BASE_DIR)
+
+        cls.bblfsh = "0.0.0.0:9432"
+        cls.language = "javascript"
+        cls.model_path = None
+
+    @classmethod
+    def tearDownClass(cls):
+        """Remove temporary directory with jquery repository."""
+        shutil.rmtree(cls.jquery_dir)
+
+    def tearDown(self):
+        logging.getLogger().removeHandler(self.log_handler)
+        self.listener.stop()
+        self.process.join()
+        self.db.close()
+
+    def setUp(self):
+        self.port = find_port()
+        self.db = tempfile.NamedTemporaryFile(dir=BASE_DIR)
+
+        command = "analyzer init --db sqlite:///%s --fs %s" % (self.db.name, FS.name)
+        with patch("sys.argv", command.split(" ")):
+            launch_analyzer()
+        command = ("analyzer run lookout.style.format --db sqlite:///%s "
+                   "--server localhost:%d --fs %s --log-level DEBUG"
+                   ) % (self.db.name, self.port, FS.name)
+        self.logs = logs = []
+
+        class ShadowHandler(logging.Handler):
+            def emit(self, record):
+                logs.append(logging.getLogger().handlers[0].formatter.format(record))
+
+        self.log_handler = ShadowHandler()
+        logging.getLogger().addHandler(self.log_handler)
+
+        def main():
+            with patch("sys.argv", command.split(" ")):
+                launch_analyzer()
+
+        sync = threading.Event()
+        block = EventListener.block
+
+        def sync_block(this):
+            self.listener = this
+            sync.set()
+            return block(this)
+
+        EventListener.block = sync_block
+        self.process = Thread(target=main)
+        try:
+            self.process.start()
+            sync.wait()
+        finally:
+            EventListener.block = block
+
+    def train(self):
+        if self.model_path is None:
+            launch_server(
```

Comment:
Can you do it in setUpClass? I see that there is an assertion but it shouldn't be much problem to remove it or move to setUp.

Reply:
I think it's not possible without a lot of copy-paste of code
1) model is trained only one time (that's why `self.model_path is None` was introduced)
2) training of the model requires `setUp` to prepare an environment
So there is no computational overhead (model is trained once) + it requires to copy whole `setUp` to train the model in `setUpModule`. I prefer to keep it as it's.

Reply:
You already have setupclass so we should not have setupmodule at all. Besides, all the code which is executed once must go under the static calls. Waiting for the fix of this, sorry, but otherwise is blocked.